### PR TITLE
feat: finish adapter implementation & properly type makeRequest

### PR DESCRIPTION
**Pull request by andipaetzold**
This PR adds [proper types](https://github.com/contentful/contentful-management.js/pull/740/files#diff-5d65b07a85ecfbaae4af51feabd68aefb18ecbf40d80f8076768882c66d01b73R165-R974) for all possible actions for `makeRequest`.

Based on these types, the type signatures of all rest endpoints are generated using [`RestEndpoint`](https://github.com/contentful/contentful-management.js/pull/740/files#diff-ba33b9cae3a835a634a247dc94fcc863d613ede2d5a471c4337143f68aed19aaR4-R23). The `RestEndpoint<...,...>` type was used to type every single endpoint. Probably this could be also done in `lib/adapters/REST/endpoints/index.ts` to avoid touching too many files and to reduce the amount of code.

With these added types, we can ensure type safety from the public methods to the internal RestAdapter endpoints.

By adding these types, many issues in the previously merged plain client were found and resolved.

### CODE REVIEW

File: lib/adapters/REST/endpoints/index.ts [hunk 1]
```diff
@@ -1,30 +1,65 @@
...
+import * as PreviewApiKey from './preview-api-keys'
+import * as Role from './role'
+import * as ScheduledAction from './scheduled-action'
+import * as Snapshot from './snapshot'
+import * as Space from './space'
+import * as SpaceMember from './space-member'
+import * as SpaceMembership from './space-membership'
+import * as Team from './team'
+import * as TeamMembership from './team-membership'
+import * as Tag from './tag'
```
Comment (shikaan) [inline lib/adapters/REST/endpoints/index.ts:25]: nit: sorting this list would make easier to spot missing entries

File: lib/adapters/REST/endpoints/index.ts [hunk 2]
```diff
@@ -1,30 +1,65 @@
...
+import * as Entry from './entry'
+import * as Environment from './environment'
+import * as EnvironmentAlias from './environment-alias'
+import * as Http from './http'
+import * as Locale from './locale'
+import * as Organization from './organization'
+import * as OrganizationInvitation from './organization-invitation'
+import * as OrganizationMembership from './organization-membership'
+import * as PersonalAccessToken from './personal-access-token'
+import * as PreviewApiKey from './preview-api-keys'
```
Comment (shikaan) [inline lib/adapters/REST/endpoints/index.ts:16]: name of this file is plural, as opposed to all the rest

File: lib/adapters/REST/endpoints/organization.ts [hunk 3]
```diff
@@ -1,14 +1,17 @@
 import type { AxiosInstance } from 'contentful-sdk-core'
-import { CollectionProp } from '../../../common-types'
+import { CollectionProp, GetOrganizationParams } from '../../../common-types'
 import { OrganizationProp } from '../../../entities/organization'
-import { GetOrganizationParams } from '../../../plain/common-types'
+import { RestEndpoint } from '../types'
 import * as raw from './raw'
 
-export const getAll = (http: AxiosInstance) => {
+export const getAll: RestEndpoint<'Organization', 'getAll'> = (http: AxiosInstance) => {
```
Comment (shikaan) [inline lib/adapters/REST/endpoints/organization.ts:7]: I would make it `getMany` too, for consistency™️. We can still export also `getAll` for backwards compatibility and get rid of it on the next major

File: lib/contentful-management.ts [hunk 4]
```diff
@@ -70,15 +70,14 @@ function createClient(
...
 
   const adapter = createAdapter(params)
-  const makeRequestWithoutUserAgent: MakeRequestWithoutUserAgent = (options) =>
-    adapter.makeRequest({
-      ...options,
-      userAgent,
-    })
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
```
Comment (shikaan) [inline lib/contentful-management.ts:75]: It's not clear to me why this is here. I'd add a comment to explain (and consider - if applicable - using `@ts-expect-error` if that's something we plan to solve)

File: lib/create-space-api.ts [hunk 5]
```diff
@@ -264,7 +259,7 @@ export default function createSpaceApi({
       return makeRequest({
         entityType: 'Webhook',
         action: 'getMany',
-        params: { spaceId: raw.sys.id },
+        params: { spaceId: raw.sys.id, query: undefined },
```
Comment (shikaan) [inline lib/create-space-api.ts:262]: Why explicitly `undefined`?

File: lib/methods/content-type.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  if (!field) {
+    return Promise.reject(
+      new Error(
+        `Tried to omitAndDeleteField on a nonexistent field, ${fieldId}, on the content type ${contentType.name}.`
+      )
+    )
+  }
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
```
Comment (shikaan) [inline lib/methods/content-type.ts:27]: Same comment as above

File: lib/plain/wrappers/wrap.ts [hunk 7]
```diff
@@ -16,19 +16,52 @@ export type MarkOptional<BaseType, Keys extends keyof BaseType = keyof BaseType>
...
   { makeRequest, defaults }: WrapParams,
-  entityType: string,
-  action: string
-) => {
-  return (params?: Params, payload?: Payload, headers?: Record<string, unknown>): Promise<any> =>
+  entityType: ET,
+  action: Action
+): WrapFn<ET, Action> => {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
```
Comment (shikaan) [inline lib/plain/wrappers/wrap.ts:50]: Same same same 😓  Rule of thumb should be to comment `@ts-ignore`s so we know why they are there

File: lib/entities/preview-api-key.ts [hunk 8]
```diff
@@ -18,12 +18,12 @@ function createPreviewApiKeyApi() {
...
 /**
  * @private
- * @param http - HTTP client instance
+ * @param makeRequest - function to make requests via an adapter
  * @param data - Raw api key data
  * @return Wrapped preview api key data
  */
 export function wrapPreviewApiKey(
-  _makeRequest: MakeRequestWithoutUserAgent,
+  _makeRequest: MakeRequest,
```
[the remark this answers is not in the document]
Comment (andipaetzold) [reply lib/entities/preview-api-key.ts:26]: The `_` is to ignore "unused variable name"-errors. `makeRequest` is not used within `wrapPreviewApiKey`.

File: lib/adapters/REST/endpoints/index.ts [hunk 9]
```diff
@@ -1,30 +1,65 @@
...
+import * as ScheduledAction from './scheduled-action'
+import * as Snapshot from './snapshot'
+import * as Space from './space'
+import * as SpaceMember from './space-member'
+import * as SpaceMembership from './space-membership'
+import * as Team from './team'
+import * as TeamMembership from './team-membership'
+import * as Tag from './tag'
+import * as TeamSpaceMembership from './team-space-membership'
+import * as UIExtension from './ui-extension'
```
[the remark this answers is not in the document]
Comment (andipaetzold) [reply lib/adapters/REST/endpoints/index.ts:27]: Internally it is almost always called "UI Extension". Only the public api uses "Extension"

[the remark this answers is not in the document]
Comment (shikaan) [reply lib/adapters/REST/endpoints/index.ts:27]: I would still unify lingo, if it's possible

**Review by shikaan — APPROVED** [review]
LGTM 🎉

File: lib/adapters/REST/endpoints/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (andipaetzold) [reply lib/adapters/REST/endpoints/index.ts:27]: I'll try :)

**Pull request merged**